packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -129,9 +173,38 @@ export class ConnectionStateHandler extends EventEmitterWithErrorHandling<IConne
                 }
             }
             this._clientId = this.pendingClientId;
+            // Set isDirty to false as this is a fresh connection.
+            this.isDirty = false;
         } else if (value === ConnectionState.Disconnected) {
             // Important as we process our own joinSession message through delta request
             this._pendingClientId = undefined;
+            // Only wait for "leave" message if we have some outstanding ops and the client was write client as
+            // server would not accept ops from read client. Also check if the timer is not already set as we
+            // could receive "Disconnected" event multiple times without getting connected and in that case we
+            // don't want to reset the timer as we still want to wait on original client which started this timer.
+            // We also check the dirty state of this connection as we only want to wait for the client leave of the
+            // client which created the ops. This helps with situation where a client disconnects immediately after
+            // getting connected without sending any ops(from previous client). In this case, we would join as write
+            // because there would be a diff between client seq number and clientSeqNumberObserved but then we don't
+            // want to wait for newly disconnected client to leave as it has not sent any ops yet.
+            if (this.handler.shouldClientJoinWrite()
+                && this.handler.client().mode === "write"
```

Comment:
I think this.handler.client().mode === "write" will always compute to false, because this is not IClient from prior connection, this is Container.client property that has some defaults (and mode is default mode this client connects to, when shouldClientJoinWrite() returns false)

Reply:
Using the right mode now.